stdlib/2/urllib2.pyi

```diff
@@ -1,18 +1,29 @@
 
 import ssl
-from typing import Any, AnyStr, Dict, List, Union, Optional, Mapping, Callable, Sequence, Tuple, Type
+from typing import Any, AnyStr, Callable, Dict, Iterator, List, Union, Optional, Mapping, Sequence, Tuple, Type
+from typing_extensions import Protocol
 from urllib import addinfourl
 from httplib import HTTPResponse
+from abc import ABCMeta
 
 _string = Union[str, unicode]
 
+class _SupportsRead(Protocol):
+    def read(self, _size: int = ...) -> _string: ...
+    def readline(self, _size: int = ...) -> _string: ...
+    def fileno(self) -> Optional[int]: ...
+    def readlines(self, _size: int = ...) -> List[str]: ...
+    def __iter__(self) -> Iterator[_string]: ...
+    def next(self) -> _string: ...
+    def close(self) -> None: ...
+
 class URLError(IOError):
     reason = ...  # type: Union[str, BaseException]
 
 class HTTPError(URLError, addinfourl):
     code = ...  # type: int
-    headers = ...  # type: Dict[str, str]
-    def __init__(self, url, code, msg, hdrs, fp) -> None: ...
+    headers: Mapping[str, str] = ...
+    def __init__(self, url, code: int, msg: str, hdrs: Mapping[str, str], fp: _SupportsRead) -> None: ...
```

Comment:
As far as I can tell this only needs read, readlines, and close, not the much bigger protocol you wrote. (I'm going off the implementation of `urllib.addbase`.)

Reply:
What about `close`? IIRC the close method of addbase uses it.

Reply:
Good catch; done in 35bbe82dc22d34c5db811e76d25233b7e4958afa